### ISSUE TITLE
Support recursive launches/dynamic parallelism

### ISSUE THREAD

**Issue by adityaatluri** [inducer/pycuda#45]
I am trying to implement dynamic parallelism in my PyCUDA file. I am getting this error message. What to do?

**Comment by chschnell** [comment]
Hello,

first, thanks for PyCuda! I've added support for dynamic parallelism (more or less as discussed above), please see the branch `jit-link` in my fork:

https://github.com/chschnell/pycuda/tree/jit-link

I added two classes:
- C++ class `Linker` plus Boost.Python wrapping (as `pycuda.driver.Linker`) in wrap_cudadrv.cpp
- Python class `pycuda.compiler.JitLinkModule` in compiler.py

The C++ class `Linker` exposes the cuLink* functions to the Python interpreter (to be exact: `cuLinkCreate`, `cuLinkAddData`, `cuLinkAddFile`, `cuLinkComplete`, `cuLinkDestroy` and `cuModuleLoadData`), which are then used by the Python class. The Python class can be used as an almost drop-in replacement for class `SourceModule` (these two classes should be united in some way, maybe a common base class, currently I just lazily copy-pasted some parts from `SourceModule`).

The cuLink* functions or not exposed 1:1 in C++. If I had exposed the cuLink* API directly to Python there would have been the concern of a possible memory leak. The reason is that `cuLinkCreate()` allocates a ressource for us that we are expected to eventually release by calling `cuLinkDestroy()`. So I decided to wrap that resource handle in `Linker` and to release it in the destructor. From there it was clear to turn the cuLink* API calls into methods of that class and so I ended up integrating it all in there.

CUDA's JIT linker supports two sources of input, either a file name (`cuLinkAddFile()`) or some raw data block in memory (`cuLinkAddData()`). They basically do the same, so you can either pass a file name to `cuLinkAddFile()`, or load that file's content into memory and pass it to `cuLinkAddData()`, it has the same effect.

The JIT linker also supports five different types of input (defined in `enum CUjitInputType`), these are important and we need those. So I've mapped that enum using Boost.Python to `pycuda.driver.jit_input_type.*` like this:
```
CU_JIT_INPUT_CUBIN =     pycuda.driver.jit_input_type.CUBIN
CU_JIT_INPUT_PTX =       pycuda.driver.jit_input_type.PTX
CU_JIT_INPUT_FATBINARY = pycuda.driver.jit_input_type.FATBINARY
CU_JIT_INPUT_OBJECT =    pycuda.driver.jit_input_type.OBJECT
CU_JIT_INPUT_LIBRARY =   pycuda.driver.jit_input_type.LIBRARY
```
A practical example. Let's say you had these typical lines in your code:
```python
from pycuda.compiler import SourceModule

mod = SourceModule(my_nvcc_src, options=['-O3'])

my_kernel_func = mod.get_function('my_kernel_func')
```
In order to switch to dynamic parallelism, you'd need to replace that code with:
```python
from pycuda.compiler import JitLinkModule
from pycuda.driver import jit_input_type

mod = JitLinkModule()
mod.add_source(my_nvcc_src, nvcc_options=['-O3', '-rdc=true', '-lcudadevrt'])
mod.add_file(os.environ['CUDA_PATH'] + '\\lib\\x64\\cudadevrt.lib', jit_input_type.LIBRARY)
mod.link()

my_kernel_func = mod.get_function('my_kernel_func')
```
Note that after invoking `mod.link()`, you'd use `mod` in the exact same way as you did before when using the `SourceModule` class. This certainly needs more documentation, but that's basicaly all that's to it from a PyCuda user's perspective.

It's tested, though not it in-depth yet, but from first attempts I've noticed a considerable performance impact by only linking the device runtime library `cudadevrt.lib` to my application, a drop of around 20% in overall performance. I'm not sure how well I can extrapolate from my tests, but I found [this](https://docs.nvidia.com/cuda/pdf/CUDA_Dynamic_Parallelism_Programming_Guide.pdf) from NVIDIA to keep in mind:

> **C.4.2.2. Dynamic-parallelism-enabled Kernel Overhead**
> 
> System software which is active when controlling dynamic launches may impose an overhead on any kernel which is running at the time, whether or not it invokes kernel launches of its own. This overhead arises from the device runtime's execution tracking and management software and may result in decreased performance for e.g., library calls when made from the device compared to from the host side. This overhead is, in general, incurred for applications that link against the device runtime library.

From the comments above not everything was clear so I had to improvise a little here and there, but it's all up for discussion. Please let me know if you want any changes, and also where to provide full documentation of the three new data types.

Cheers,
Christian

**Comment by inducer** [comment]
https://gitlab.tiker.net/inducer/pycuda/merge_requests/2

**Comment by isaactian** [comment]
Does the package downloaded from the official installation instructions now support dynamic parallelism? Is this resolved?

https://pypi.python.org/pypi/pycuda

**Comment by jllanfranchi** [comment]
Presumably not yet, @isaactian since the last upload to pypi for PyCUDA was on date "2016-07-26," well before the above code was merged into master in the gitlab repo.

**Comment by inducer** [comment]
I've just released 2017.1, which contains this work.

**Comment by isaactian** [comment]
If I follow Christian's instructions for jitlinkmodule, python says the module does not exist (browsing the source confirms this).

If I just set everything as normal (assuming it was silently integrated into SourceModule) and try to call a kernel from a global function I get:

error: kernel launch from __device__ or __global__ functions requires separate compilation mode.

How do I get around this? Thanks.
